miner/src/pool/queue.rs

```diff
@@ -0,0 +1,440 @@
+// Copyright 2015-2017 Parity Technologies (UK) Ltd.
+// This file is part of Parity.
+
+// Parity is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Ethereum Transaction Queue
+
+use std::{cmp, fmt};
+use std::sync::Arc;
+use std::sync::atomic::{self, AtomicUsize};
+use std::collections::BTreeMap;
+
+use ethereum_types::{H256, U256, Address};
+use parking_lot::RwLock;
+use rayon::prelude::*;
+use transaction;
+use txpool::{self, Verifier};
+
+use pool::{self, scoring, verifier, client, ready, listener, PrioritizationStrategy};
+use pool::local_transactions::LocalTransactionsList;
+
+type Listener = (LocalTransactionsList, (listener::Notifier, listener::Logger));
+type Pool = txpool::Pool<pool::VerifiedTransaction, scoring::NonceAndGasPrice, Listener>;
+
+/// Max cache time in milliseconds for pending transactions.
+///
+/// Pending transactions are cached and will only be computed again
+/// if last cache has been created earler than `TIMESTAMP_CACHE` ms ago.
+/// This timeout applies only if there are local pending transactions
+/// since it only affects transaction Condition.
+const TIMESTAMP_CACHE: u64 = 1000;
+
+/// Transaction queue status.
+#[derive(Debug, Clone, PartialEq)]
+pub struct Status {
+	/// Verifier options.
+	pub options: verifier::Options,
+	/// Current status of the transaction pool.
+	pub status: txpool::LightStatus,
+	/// Current limits of the transaction pool.
+	pub limits: txpool::Options,
+}
+
+impl fmt::Display for Status {
+	fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+		writeln!(
+			fmt,
+			"Pool: {current}/{max} ({senders} senders; {mem}/{mem_max} kB) [minGasPrice: {gp} Mwei, maxGas: {max_gas}]",
+			current = self.status.transaction_count,
+			max = self.limits.max_count,
+			senders = self.status.senders,
+			mem = self.status.mem_usage / 1024,
+			mem_max = self.limits.max_mem_usage / 1024,
+			gp = self.options.minimal_gas_price / 1_000_000.into(),
+			max_gas = cmp::min(self.options.block_gas_limit, self.options.tx_gas_limit),
+		)
+	}
+}
+
+#[derive(Debug)]
+struct CachedPending {
+	block_number: u64,
+	current_timestamp: u64,
+	nonce_cap: Option<U256>,
+	has_local_pending: bool,
+	pending: Option<Vec<Arc<pool::VerifiedTransaction>>>,
```

Comment:
why is it `Option<Vec<_>>` ? It could be only `Vec<_>`

Reply:
is `none()` being created often?

Reply:
Yeah, `Option` does distinguish between: no-cached pending set and cached empty pending set. We could try to construct pending set every time we see an empty vector, since it should be cheap anyway. Is that what you mean?